components/omega/doc/design/Config.md

```diff
@@ -0,0 +1,332 @@
+### 2.2 Requirement: Standard format
+
+Configuration files must conform to a standard format for ease in parsing
+and for potential interoperability. Examples of standards typically used
+include YAML, JSON, XML.
+
+### 2.3 Requirement: Archiving and provenance
+
+The entire model configuration must be able to be saved and archived to
+act as model provenance. 
+
+### 2.4 Requirement: Internal accessibility
+
+Although most input parameters will only be relevant to a particular
+module or class, it is inevitable that some parameterizations or process
+models will have dependencies on other configuration choices. All
+configuration variables must therefore be accessible to all other
+modules/classes.
+
+### 2.5 Requirement: Support of data types
+
+Model configuration must support parameters in the standard data types,
+including logical (boolean), integer, float/double, and string data types.
+Limited support for vectors or arrays of the above types may be desirable.
+
+### 2.6 Requirement: Efficiency and parallelism
+
+Configuration should never consume significant resources, in either run
+time or storage. For parallel execution, the configuration
+will need to be replicated across MPI ranks, though in many cases, the
+variables will be manipulated into a different form (eg string variables
+describing choices will be converted to enums). This may happen before
+or after replication.
+
+### 2.8 Desired: Single configuration input
+
+For simplicity, we desire a single file for model configuration where users
+can define all model configuration details and provenance can be maintained
+more easily. If multiple files are required, there should be obvious links
+or includes in the main configuration file to make the dependency clear.
+
+### 2.9 Required: Hierarchy or grouping
+
+For both readability and for easier encapsulation by modules/classes, the 
+configuration should be organized in a logical way to make it clear where
+specific configuration variables are to be used. Parameters primarily or
+specifically associated with a module or class should be in a grouping
+of those parameters with the configuration.
+
+### 2.10 Requirement: Language support
+
+Because the configuration will also be used for provenance, it is likely
+that the configuration input file will need to be read by other languages
+outside of the OMEGA C++ model. This requires an ability to parse a 
+configuration input file from other common languages (eg python).
+
+### 2.11 Requirement: Optional or missing values
+
+If a configuration variable is missing from the configuration input file,
+there must be an option to either supply a default value or throw an error
+and exit, depending on the user choice. If a default value is supplied,
+then the default value must be added to the configuration so that it is
+included in the output and become part of the provenance.
+
+### 2.12 Requirement: Extra values
+
+If extra or unexpected values are encountered, they will be ignored.
```

Comment:
In compass, we have found that this leads to unexpected behavior.  For example, there may be a typo in a config option that is then ignored and the default is used instead.  Or a config option may be renamed but the compass value is still being used.

It would therefore would be desirable to at least print warnings for config options that are not used.

Reply:
Reading through the implementation, this may be difficult to fulfill, since it would require somehow marking each config variable as accessed and, only after access is complete, determining which config variables were missed.  This is something we will do on the polaris side as we have in compass but it may not make sense to do on the Omega side.

Reply:
I think this was worded poorly, but it refers to the case where there is a key/value pair in the config file that is never used by the code. That would typically happen if some option was removed but the config file still thinks it's there. There is really no way to check for this (similar to stale registry entries) other than to have the code build an internal config based on what it's expecting and compare the input and internally generated versions. 

In the case of a typo, the bad entry would be ignored, but presumably the expected entry would be missing and that would generate an error (see other comments below).

@cbegeman on the `possible_values`, I hope these show up in three places - internally in the init routine, in the user/ref docs, and we could add them as comments in a reference omega.yml file.  Which brings up the question of an added step to keep this reference yaml file up to date together with the documentation.

Reply:
It seems like it will be important for outside testing (e.g. with polaris) if there is a way to pretty definitively keep track of both the default values and possible values outside of the Omega code itself.  This could allow us to check for bad input config options (e.g. old ones that aren't valid anymore or typos) and to ensure that tests cover all the allowed config settings wherever possible.  @philipwjones could you envision Omega writing out one or more files when it's built, a bit like MPAS components currently do, that provides at least the full set of config options with defaults, and maybe even a list of possible values?  Parsing these out of c++ code or documentation seems pretty daunting.  Getting them out of comments in a reference `omega.yml` could work if the syntax is pretty standardized.

Reply:
@philipwjones Right, it gets more difficult to maintain the more places there is redundant information. Could user/ref docs pull from the comments in the yaml file? And then we'll standardize the `possible_values` argument as @xylar suggested so that we can run a check in polaris before getting to model init?

Reply:
I'll have to think about the `possible_values` stuff some more and how to generate a default yaml input file. I don't want to go back to registry so just need to think of a way to extract from source or header files with some standardized form.

Reply:
And add as an addition requirement....